File: core/src/main/javascript/io/JsonMslTokenizer.js
/**
 * Copyright (c) 2015-2018 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * <p>Create a new {@link MslTokenizer} that parses JSON-encoded MSL
 * messages.</p>
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
(function(require, module) {
	"use strict";
	
	var MslTokenizer = require('../io/MslTokenizer.js');
	var AsyncExecutor = require('../util/AsyncExecutor.js');
	var InterruptibleExecutor = require('../util/InterruptibleExecutor.js');
	var MslEncoderException = require('../io/MslEncoderException.js');
	var ClarinetParser = require('../io/ClarinetParser.js');
	var JsonMslObject = require('../io/JsonMslObject.js');
	var MslConstants = require('../MslConstants.js');
	var TextEncoding = require('../util/TextEncoding.js');
	
	/**
	 * Source input stream read size in bytes (64KiB). The larger the read
	 * size, the shorter the recursive call stack.
	 * 
	 * @const
	 * @type {number}
	 */
	var READ_SIZE = 64 * 1024;

    /**
     * Maximum JSON value size in characters (10MiB).
     *
     * @const
     * @type {number}
     */
    var MAX_CHARACTERS = 10 * 1024 * 1024;
    
    /**
     * Closing curly brace character code.
     * 
     * @const
     * @type {number}
     */
    var CLOSING_BRACE = 125;
    
    var JsonMslTokenizer = module.exports = MslTokenizer.extend({
        /**
         * <p>Create a new JSON MSL tokenizer that will read tokens off the
         * provided input stream.</p>
         * 
         * @param {MslEncoderFactory} encoder MSL encoder factory.
         * @param {InputStream} source JSON input stream.
         */
        init: function init(encoder, source) {
            init.base.call(this);
            
            // The properties.
            var props = {
                /** @type {MslEncoderFactory} */
                _encoder: { value: encoder, writable: false, enumerable: false, configurable: false },
                /** @type {InputStream} */
                _source: { value: source, writable: false, enumerable: false, configurable: false },
                /** @type {string} */
                _charset: { value: TextEncoding.Encoding.UTF_8, writable: false, enumerable: false, configurable: false },
                /** @type {string} */
                _lastJson: { value: 0, writable: true, enumerable: false, configurable: false },
                /** @type {ClarinetParser} */
                _parser: { value: new ClarinetParser(), writable: false, enumerable: false, configurable: false },
            };
            Object.defineProperties(this, props);
        },
        
        /** @inheritDoc */
        close: function close(timeout, callback) {
            var self = this;
            
            InterruptibleExecutor(callback, function() {
                // If there is no unparsed data then there is nothing to clean
                // up.
                if (this._parser.unparsedCount() == 0)
                    return true;
                
                // Otherwise reset the source input stream and skip bytes equal
                // to the parsed data length in encoded bytes. This will allow
                // the input stream to be used again by a different MSL
                // tokenizer.
                //
                // If we did the right thing in this class' other functions,
                // calling reset() should not throw an exception.
                var parsedCount = this._lastJson.length - this._parser.unparsedCount();
                var parsedJson = this._lastJson.substring(0, parsedCount);
                var encodedData = TextEncoding.getBytes(parsedJson, this._charset);
                this._source.reset();
                this._source.skip(encodedData.length, timeout, {
                    result: function(count) {
                        AsyncExecutor(callback, function() {
                            if (count != encodedData.length)
                                throw new MslEncoderException("Only skipped " + count + " of " + encodedData.length + " bytes. Source input stream may not be reusable for additional MSL messages.");
                            return true;
                        }, self);
                    },
                    timeout: function(count) {
                        AsyncExecutor(callback, function() {
                            if (count != encodedData.length)
                                throw new MslEncoderException("Only skipped " + count + " of " + encodedData.length + " bytes. Source input stream may not be reusable for additional MSL messages.");
                            return true;
                        }, self);
                    },
                    error: callback.error,
                });
            }, self);
        },
        
        /** @inheritDoc */
        abort: function abort() {
            abort.base.call(this);
            this._source.abort();
        },
        
        /**
         * <p>Feed more data from the source input stream into the parser until
         * another value has been found. This value is returned.</p>
         * 
         * @param {number} read timeout in milliseconds.
         * @param {{result: function(*), timeout: function(), error: function(Error)}}
         *        callback the callback that will receive the next parsed
         *        value, be notified of timeout, or any thrown exceptions.
         * @throws MslEncoderException if a JSON object exceeds the maximum
         *         permitted size or there is an error reading from the source
         *         input stream.
         */
        parseData: function parseData(timeout, callback) {
            var self = this;
            
            InterruptibleExecutor(callback, function() {
                // Mark the source input stream. We will need to return to this
                // position when we suceed in parsing the JSON.
                this._source.mark(READ_SIZE);
                
                // Read the next chunk of data.
                this._source.read(READ_SIZE, timeout, {
                    result: function(data) {
                        InterruptibleExecutor(callback, function() {
                            // On end of stream return null for the value.
                            if (!data) return null;
        
                            // Aborted responses send valid but empty data, treat
                            // it as end of stream.
                            if (!data.length) return null;
                            
                            // Convert the collected bytes to a string.
                            this._lastJson = TextEncoding.getString(data, this._charset);
                            
                            // If the new unparsed data size exceeds the maximum
                            // allowed then error.
                            var unparsedCount = this._parser.unparsedCount() + this._lastJson.length;
                            if (unparsedCount > MAX_CHARACTERS)
                                throw new MslEncoderException("JSON parsing stopped after reaching " + unparsedCount + " unparsed characters.");
                            
                            // Attempt to parse the JSON.
                            this._parser.write(this._lastJson);
    
                            // If we got something then return the value.
                            var value = this._parser.nextValue();
                            if (value !== undefined)
                                return value;
        
                            // Otherwise retry. This will discard the current
                            // mark position which is okay as we definitely
                            // need to read past the read limit.
                            //
                            // Use setTimeout to avoid blowing the stack, which
                            // is also faster.
                            setTimeout(function() {
                                self.parseData(timeout, callback);
                            }, 0);
                        }, self);
                    },
                    timeout: callback.timeout,
                    error: function(e) {
                        AsyncExecutor(callback, function() {
                            throw new MslEncoderException("Error reading from the source input stream.", e);
                        }, self);
                    },
                });
            }, self);
        },
        
        /** @inheritDoc */
        next: function next(timeout, callback) {
            var self = this;
            
            InterruptibleExecutor(callback, function() {
                // Ask the parser for the next value.
                var value = this._parser.nextValue();
                
                // If we received a value, wrap it.
                if (value !== undefined) {
                    wrapValue(value);
                    return;
                }

                // Otherwise try to parse more data.
                this.parseData(this._timeout, {
                    result: function(value) {
                        InterruptibleExecutor(callback, function() {
                            // If aborted then return null.
                            if (this._aborted)
                                return null;

                            // If we've reached the end of the stream then
                            // return null.
                            if (!value)
                                return null;

                            // Wrap the value.
                            wrapValue(value);
                        }, self);
                    },
                    timeout: callback.timeout,
                    error: callback.error,
                });
            }, self);
            
            function wrapValue(value) {
                InterruptibleExecutor(callback, function() {
                    // Error if the value is not an object.
                    if (typeof value !== 'object')
                        throw new MslEncoderException("Malformed MSL message. Parsed " + typeof value + " instead of object.");
                    return new JsonMslObject(this._encoder, value);
                }, self);
            }
        },
    });
})(require, (typeof module !== 'undefined') ? module : mkmodule('JsonMslTokenizer'));

File: core/src/main/cpp/io/JsonMslTokenizer.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <io/JsonHandler.h>
#include <io/JsonMslObject.h>
#include <io/JsonMslTokenizer.h>
#include <rapidjson/document.h>
#include <rapidjson/error/en.h>
#include <rapidjson/stringbuffer.h>
#include <rapidjson/writer.h>
#include <iostream>
#include <sstream>

//#define DEBUG
#include <util/Debug.h>
#include <memory>

#define TIMEOUT 60000

using namespace rapidjson;
using namespace std;

namespace netflix {
namespace msl {
namespace io {

namespace {
#if DEBUG_JSON
util::Debug myCout(cout);
#else
    std::ostream os(NULL);
    util::Debug myCout(os);
#endif
}

MslStreamWrapper::Ch MslStreamWrapper::Peek() const
{
	if (cache_.size() > 0)
		return cache_[0];
	ByteArray buf(1);
	int count = stream_->read(buf, 0, 1, TIMEOUT);
	if (RAPIDJSON_LIKELY(count != -1)) {
		cache_.push_back(buf[0]);
		return buf[0];
	}
	return '\0';
}

MslStreamWrapper::Ch MslStreamWrapper::Take()
{
	if (cache_.size() > 0) {
		Ch c = cache_.front();
		cache_.erase(cache_.begin());
		++count_;
		return c;
	}
	ByteArray buf(1);
	int count = stream_->read(buf, 0, 1, TIMEOUT);
	if (RAPIDJSON_LIKELY(count != -1)) {
		++count_;
		return buf[0];
	}
	return '\0';
}

const MslStreamWrapper::Ch* MslStreamWrapper::Peek4() const
{
	int len = static_cast<int>(4 - cache_.size());
	if (len > 0) {
		ByteArray buf(static_cast<size_t>(len));
		int count = stream_->read(buf, 0, static_cast<size_t>(len), TIMEOUT);
		for (int i = 0; i < count; ++i)
			cache_.push_back(buf[static_cast<size_t>(i)]);
	}
	if (cache_.size() >= 4)
		return &cache_[0];
	return 0;
}

shared_ptr<MslObject> JsonMslTokenizer::next(int /*timeout*/)
{
    // Parse a JSON object to a DOM
    Document document;
    const int parseFlags = kParseStopWhenDoneFlag;
    try {
        if (document.ParseStream<parseFlags>(mslstreamwrapper_).HasParseError())
        {
            const ParseErrorCode e = document.GetParseError();
            if (e == kParseErrorDocumentEmpty) // no more, return empty
                return shared_ptr<MslObject>();

            // We found some other parse error.
            std::stringstream ss;
            ss << "Invalid JSON encoding: " << GetParseError_En(e);
            ss << " at offset " << document.GetErrorOffset();

            // So we don't keep seeing the same error, advance the stream
            // forward, discarding everything until we find a '{'.
            uint8_t b = mslstreamwrapper_.Peek();
            while (b && b != '{')
            {
            	mslstreamwrapper_.Take(); // discard current char
                b = mslstreamwrapper_.Peek(); // peek at next, returns null at eos
            }

            throw MslEncoderException(ss.str());
        }
    }
    catch (const MslEncoderException& e) {
        throw e;
    }
    catch (...) {
        throw MslEncoderException("Invalid JSON encoding");
    }

    StringBuffer s;
    Writer<StringBuffer> writer(s);
    document.Accept(writer);
    myCout << "### " << s.GetString() << endl;

    // We now have a well-formed DOM. Translate it to a MslObject.
    shared_ptr<MslObject> mslObject = make_shared<JsonMslObject>();
    JsonHandler handler(mslObject);
    try {
        document.Accept(handler);
    }
    catch (const Exception& e) {
        throw MslEncoderException("Error when building JsonMslObject", e);
    }
    catch (...) {
        throw MslEncoderException("Unknown error when building JsonMslObject");
    }

    myCout << "$$$ " << mslObject->toString() << endl;

    return mslObject;
}

} /* namespace io */
} /* namespace msl */
} /* namespace netflix */
